pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -31,48 +31,82 @@ func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashb
-		panelUid := rule.Annotations[ngmodels.PanelIDAnnotation]
+type itemWithMetadata struct {
+	annotations.Item
+	dashUID       string
+	parsedPanelID int64
+}
 
-		panelId, err := strconv.ParseInt(panelUid, 10, 64)
-		if err != nil {
-			h.log.Error("error parsing panelUID for alert annotation", "panelUID", panelUid, "alertRuleUID", rule.UID, "err", err.Error())
-			return
-		}
+func (h *AnnotationStateHistorian) buildAnnotations(states []state.ContextualState) []itemWithMetadata {
+	items := make([]itemWithMetadata, 0, len(states))
+	for _, state := range states {
+		h.log.Debug("alert state changed creating annotation", "alertRuleUID", state.AlertRuleUID, "newState", state.Formatted(), "oldState", state.PreviousFormatted())
 
-		query := &models.GetDashboardQuery{
-			Uid:   dashUid,
-			OrgId: rule.OrgID,
+		labels := removePrivateLabels(state.Labels)
+		annotationText := fmt.Sprintf("%s {%s} - %s", state.Rule.Title, labels.String(), state.Formatted())
+
+		item := itemWithMetadata{
+			Item: annotations.Item{
+				AlertId:   state.Rule.ID,
+				OrgId:     state.OrgID,
+				PrevState: state.PreviousFormatted(),
+				NewState:  state.Formatted(),
+				Text:      annotationText,
+				Epoch:     state.LastEvaluationTime.UnixNano() / int64(time.Millisecond),
+			},
 		}
 
-		err = h.dashboards.GetDashboard(ctx, query)
-		if err != nil {
-			h.log.Error("error getting dashboard for alert annotation", "dashboardUID", dashUid, "alertRuleUID", rule.UID, "err", err.Error())
-			return
+		dashUID, ok := state.Annotations[ngmodels.DashboardUIDAnnotation]
+		if ok {
+			panelUID := state.Annotations[ngmodels.PanelIDAnnotation]
+
+			panelID, err := strconv.ParseInt(panelUID, 10, 64)
+			if err != nil {
+				h.log.Error("error parsing panelUID for alert annotation", "panelUID", panelUID, "alertRuleUID", state.AlertRuleUID, "err", err.Error())
+				continue
+			}
+			item.dashUID = dashUID
+			item.parsedPanelID = panelID
 		}
+		items = append(items, item)
+	}
+	return items
+}
+
+func (h *AnnotationStateHistorian) recordAnnotationsSync(ctx context.Context, items []itemWithMetadata) {
+	annotations := make([]annotations.Item, 0, len(items))
+
+	for _, item := range items {
+		if item.dashUID != "" {
+			query := &models.GetDashboardQuery{
+				Uid:   item.dashUID,
+				OrgId: item.Item.OrgId,
+			}
 
-		item.PanelId = panelId
-		item.DashboardId = query.Result.Id
+			err := h.dashboards.GetDashboard(ctx, query)
+			if err != nil {
+				h.log.Error("error getting dashboard for alert annotation", "dashboardUID", item.dashUID, "alertRuleID", item.AlertId, "err", err.Error())
+				return
+			}
+
+			item.Item.PanelId = item.parsedPanelID
+			item.Item.DashboardId = query.Result.Id
+		}
+		annotations = append(annotations, item.Item)
 	}
 
-	if err := h.annotations.Save(ctx, item); err != nil {
-		h.log.Error("error saving alert annotation", "alertRuleUID", rule.UID, "err", err.Error())
+	if err := h.annotations.SaveMany(ctx, annotations); err != nil {
```

Comment:
I think we need to batch here or downstream, like save no more than 1k items

Reply:
I think we should first write a utility in sqlstore for batched multi-insert, with a default of 1k.

We should use it as our main API for multi-insert rather than the ORM directly. This should be used in the migration, in the batched history records, and also here.

I'm honestly shocked that the ORM does not do this for us. Every major ORM i've ever used does this transparently. Let's make a solution just above that level rather than deal with it here.

Reply:
I agree with you Alex, but I think we still need an upper bound before we merge this change - we’ve seen big batches of alert instances lock up the system already.

lf there’s a feature toggle on the batch inserts we can have a safe fallback with the old code and write down a TODO for batch inserts in sqlstore, or we can write the sqlstore changes first.

Reply:
Created a utility to enforce this in a separate PR:
https://github.com/grafana/grafana/pull/57632

Reply:
This PR is now updated to use the batching utility.

The batch size changes based on which dialect is being used.